velox/functions/prestosql/ArrayAllMatch.cpp

```diff
@@ -0,0 +1,135 @@
+ * limitations under the License.
+ */
+
+#include "velox/expression/EvalCtx.h"
+#include "velox/expression/Expr.h"
+#include "velox/expression/ScopedVarSetter.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/functions/lib/LambdaFunctionUtil.h"
+#include "velox/functions/lib/RowsTranslationUtil.h"
+#include "velox/vector/FunctionVector.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+class AllMatchFunction : public exec::VectorFunction {
+ public:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    exec::LocalDecodedVector arrayDecoder(context, *args[0], rows);
+    auto& decodedArray = *arrayDecoder.get();
+
+    auto flatArray = flattenArray(rows, args[0], decodedArray);
+    auto offsets = flatArray->rawOffsets();
+    auto sizes = flatArray->rawSizes();
+
+    std::vector<VectorPtr> lambdaArgs = {flatArray->elements()};
+    auto numElements = flatArray->elements()->size();
+
+    SelectivityVector finalSelection;
+    if (!context.isFinalSelection()) {
+      finalSelection = toElementRows<ArrayVector>(
+          numElements, *context.finalSelection(), flatArray.get());
+    }
+
+    VectorPtr matchBits;
+    auto elementToTopLevelRows = getElementToTopLevelRows(
+        numElements, rows, flatArray.get(), context.pool());
+
+    // Loop over lambda functions and apply these to elements of the base array,
+    // in most cases there will be only one function and the loop will run once.
+    context.ensureWritable(rows, BOOLEAN(), result);
+    auto flatResult = result->asFlatVector<bool>();
+    exec::LocalDecodedVector bitsDecoder(context);
+    auto it = args[1]->asUnchecked<FunctionVector>()->iterator(&rows);
+
+    // Turn off the ThrowOnError flag and reset errors in the
+    // context to nullptr, so we only handle errors in this function itself.
+    auto topLevelThrowOnError = context.throwOnError();
+    ScopedVarSetter throwOnError(context.mutableThrowOnError(), false);
+    while (auto entry = it.next()) {
+      auto elementRows =
+          toElementRows<ArrayVector>(numElements, *entry.rows, flatArray.get());
+      auto wrapCapture = toWrapCapture<ArrayVector>(
+          numElements, entry.callable, *entry.rows, flatArray);
+      entry.callable->apply(
+          elementRows,
+          finalSelection,
+          wrapCapture,
+          &context,
+          lambdaArgs,
+          elementToTopLevelRows,
+          &matchBits);
+
+      bitsDecoder.get()->decode(*matchBits, elementRows);
+      entry.rows->applyToSelected([&](vector_size_t row) {
+        auto size = sizes[row];
+        auto offset = offsets[row];
+        auto allMatch = true;
+        auto hasNull = false;
+
+        for (auto i = 0; i < size; ++i) {
+          auto idx = offset + i;
+          if (bitsDecoder->isNullAt(idx)) {
```

Comment:
Here, we need to know whether 'idx' has value or error. If it has an error we cannot trust the value.

Reply:
For each element's result of a row, it would be null if this element has an error or the element is null itself. So the return value of this row depends on other elements' results. It would return false if other elements' have false. If no false on other elements' results, it would return null when no error on this element or throw an exception when error on this element.

The first round is to mark hasNull and allMatch flag, and the second round(error handling) is to decide whether setError or return null/true/false.

```C++
if null at element then // null itself or error
  if no false elements
     if has an error then
          setError
     else
          return null
  else
    return false
```

Reply:
> For each element's result of a row, it would be null if this element has an error 

I'm not sure that's correct. Why do you think it is?

Reply:
I believe for element rows with errors, the result value can be anything. It is not necessarily null.

Reply:
>I'm not sure that's correct. Why do you think it is?
I debug the test case :)

>I believe for element rows with errors, the result value can be anything. It is not necessarily null.
OK, I will investigate it by reading the code (maybe Expr::evalAllImpl?).

Reply:
This is something that's difficult to see by reading the case, but there is nothing that requires an expression to fill in a null for a row that reports an error. TRY expression explicitly populates such rows with nulls, but if there is no TRY there is no guarantee about the values in rows with errors.

Reply:
@mbasmanova I've added `EvalCtx.errorMask_` to track the error of each element and use it to skip the untrust value in the `ArrayAllMatch`. Do you think this is reasonable?